File: qgen/pdf_splitter.py
from __future__ import annotations

from pathlib import Path

from pypdf import PdfReader

from qgen.models import Segment


def extract_page_texts(pdf_path: str | Path) -> list[str]:
    reader = PdfReader(str(pdf_path))
    texts: list[str] = []
    for page in reader.pages:
        texts.append((page.extract_text() or "").strip())
    return texts


def build_segments_from_page_texts(
    source_pdf: str,
    page_texts: list[str],
    pages_per_segment: int,
) -> list[Segment]:
    segments: list[Segment] = []
    if pages_per_segment <= 0:
        raise ValueError("pages_per_segment must be > 0")

    for start in range(0, len(page_texts), pages_per_segment):
        end = min(start + pages_per_segment, len(page_texts))
        chunk_pages = page_texts[start:end]
        merged_text = "\n\n".join([t for t in chunk_pages if t.strip()]).strip()
        segments.append(
            Segment(
                source_pdf=source_pdf,
                segment_index=len(segments),
                page_start=start + 1,
                page_end=end,
                text=merged_text,
            )
        )
    return segments


def split_pdf_into_segments(
    pdf_path: str | Path,
    pages_per_segment: int,
) -> list[Segment]:
    path = Path(pdf_path)
    page_texts = extract_page_texts(path)
    return build_segments_from_page_texts(
        source_pdf=path.name,
        page_texts=page_texts,
        pages_per_segment=pages_per_segment,
    )

File: web/lib/pdfSplitter.ts
import type { Segment } from "./types";

// Dynamic import so Next.js can tree-shake the client bundle and so the worker
// entry is only required at runtime on the Node side of the API route.
async function loadPdfjs() {
  // Legacy build is the Node-friendly distribution.
  const pdfjs = await import("pdfjs-dist/legacy/build/pdf.mjs");
  // Disable worker: run on the Node thread.
  // @ts-expect-error - GlobalWorkerOptions exists at runtime
  pdfjs.GlobalWorkerOptions.workerSrc = false;
  return pdfjs;
}

export async function extractPageTexts(buffer: Uint8Array): Promise<string[]> {
  const pdfjs = await loadPdfjs();
  const loadingTask = pdfjs.getDocument({
    data: buffer,
    disableFontFace: true,
    useSystemFonts: false,
    isEvalSupported: false,
  });
  const doc = await loadingTask.promise;
  const pages: string[] = [];
  for (let i = 1; i <= doc.numPages; i += 1) {
    const page = await doc.getPage(i);
    const content = await page.getTextContent();
    const text = content.items
      .map((item) =>
        "str" in item && typeof item.str === "string" ? item.str : ""
      )
      .join(" ")
      .replace(/\s+/g, " ")
      .trim();
    pages.push(text);
  }
  await doc.destroy();
  return pages;
}

export function buildSegmentsFromPageTexts(
  sourcePdf: string,
  pageTexts: string[],
  pagesPerSegment: number
): Segment[] {
  if (pagesPerSegment <= 0) {
    throw new Error("pagesPerSegment must be > 0");
  }
  const segments: Segment[] = [];
  for (let start = 0; start < pageTexts.length; start += pagesPerSegment) {
    const end = Math.min(start + pagesPerSegment, pageTexts.length);
    const chunk = pageTexts.slice(start, end).filter((t) => t.trim().length > 0);
    const merged = chunk.join("\n\n").trim();
    segments.push({
      sourcePdf,
      segmentIndex: segments.length,
      pageStart: start + 1,
      pageEnd: end,
      text: merged,
    });
  }
  return segments;
}

export async function splitPdfIntoSegments(
  sourcePdf: string,
  buffer: Uint8Array,
  pagesPerSegment: number
): Promise<{ segments: Segment[]; pageCount: number }> {
  const pageTexts = await extractPageTexts(buffer);
  const segments = buildSegmentsFromPageTexts(sourcePdf, pageTexts, pagesPerSegment);
  return { segments, pageCount: pageTexts.length };
}
